File: lib/pipeline/downsample.ts
import { linearToSrgb, oklabToRgbGamutMapped, rgbToOklab, SRGB_TO_LINEAR } from "../color/color";
import type { CellColorBuffer, PixelBuffer, RGB } from "../types";

export interface GridDimensions {
  width: number;
  height: number;
}

/** Grid size for a source image given the longer side's stitch count, rounded so a fractional input can't reach a typed-array allocation (code review 2026-09-09, finding 8). */
export function gridDimensionsFor(
  sourceWidth: number,
  sourceHeight: number,
  longerSideStitches: number
): GridDimensions {
  const longerSide = Math.max(1, Math.round(longerSideStitches));
  if (sourceWidth >= sourceHeight) {
    const height = Math.max(1, Math.round((longerSide * sourceHeight) / sourceWidth));
    return { width: longerSide, height };
  }
  const width = Math.max(1, Math.round((longerSide * sourceWidth) / sourceHeight));
  return { width, height: longerSide };
}

/** A downsample that also reports how much of each cell the photo actually covers (G-050). */
export interface DownsampledCells {
  cells: CellColorBuffer;
  /** Per cell, the alpha-weighted share of its footprint that is opaque: 1 fully covered, 0 fully transparent. */
  coverage: Float32Array;
}

/**
 * A cell covered less than this becomes an empty stitch rather than a colour (Owner, 2026-09-20): an anti-aliased edge
 * resolves the way a stitcher would decide it, and a subject on transparency gets a clean outline instead of a fringe.
 */
export const MIN_CELL_COVERAGE = 0.5;

/** `downsampleToGridWithCoverage`, keeping only the colours — every caller that has no use for coverage. */
export function downsampleToGrid(imageData: PixelBuffer, gridWidth: number, gridHeight: number, vividTopShare = 0): CellColorBuffer {
  return downsampleToGridWithCoverage(imageData, gridWidth, gridHeight, vividTopShare).cells;
}

/**
 * Vivid (G-061): the share of each cell's footprint, by weight, whose chroma the cell keeps. Chosen by measurement
 * over both of the Owner's photos and every fixture (`docs/reviews/2026-09-22-vivid.md`); 0 is off, and the cell is
 * then exactly the area mean it has always been.
 */
export const VIVID_TOP_SHARE = 0.25;

/**
 * Vivid only rescues colour that is *inside* a stitch, so it needs a stitch to hold enough pixels for "its most
 * colourful quarter" to be a real subpopulation rather than a lucky pixel. Below this many source pixels per cell it
 * stands down and the cell is its plain area mean (D211). Measured: on fixtures at about 4 pixels a cell the chroma
 * a cell gains from noise alone reaches 0.127, larger than the 0.03-0.04 that real sub-stitch colour produces on a
 * photo at 144-400 pixels a cell, so no threshold on the gain itself can separate the two.
 */
export const VIVID_MIN_PIXELS_PER_CELL = 24;

/** Whether Vivid can act on this photo at this chart size, which is what a pattern records rather than what was asked (D211). */
export function vividApplies(sourceWidth: number, sourceHeight: number, gridWidth: number, gridHeight: number): boolean {
  return (sourceWidth * sourceHeight) / (gridWidth * gridHeight) >= VIVID_MIN_PIXELS_PER_CELL;
}

/** `max − min` of the sRGB bytes: a whole-number stand-in for chroma, so the pixels a cell keeps are picked by integer comparison in both languages rather than by a float that could round differently. */
function chromaProxy(data: Uint8ClampedArray, pixelIndex: number): number {
  const r = data[pixelIndex];
  const g = data[pixelIndex + 1];
  const b = data[pixelIndex + 2];
  return Math.max(r, g, b) - Math.min(r, g, b);
}

/**
 * Which cells are too transparent to stitch, or null when every cell is covered — the null is what keeps a photo
 * without transparency on exactly the code path it had before (G-050, criterion 3).
 */
export function emptyCellMask(coverage: Float32Array): Uint8Array | null {
  let mask: Uint8Array | null = null;
  for (let i = 0; i < coverage.length; i++) {
    if (coverage[i] >= MIN_CELL_COVERAGE) continue;
    mask ??= new Uint8Array(coverage.length);
    mask[i] = 1;
  }
  return mask;
}

/**
 * Area-weighted box downsample to one averaged RGB per cell: every source pixel contributes by its exact fractional
 * overlap with the cell's rectangle, so non-integral ratios carry no spatial bias and upscaling needs no separate path
 * (code review 2026-09-09, finding 2). Averaging is alpha-weighted, so transparent pixels don't pull colors toward
 * black, and happens in linear light: a 50/50 black/white split averages to sRGB ~188, not ~128 (D7). A cell whose
 * whole footprint is transparent becomes white.
 */
export function downsampleToGridWithCoverage(
  imageData: PixelBuffer,
  gridWidth: number,
  gridHeight: number,
  /**
   * Vivid (G-061): 0 keeps this function exactly as it was. Above 0, the cell keeps its area-mean lightness but takes
   * the chroma of its most colourful `vividTopShare` of weight, so a saturated minority inside one stitch — a red
   * print, a pink petal against cream — is not averaged into a neutral. A cell whose pixels are all equally colourful
   * selects all of them, which makes it its own mean again, so a flat region is untouched either way (D211).
   */
  vividTopShare = 0
): DownsampledCells {
  const { width: srcW, height: srcH, data } = imageData;
  const out = new Uint8ClampedArray(gridWidth * gridHeight * 3);
  const coverage = new Float32Array(gridWidth * gridHeight);
  // A whole-image decision, not a per-cell one: the ratio is the same for every cell up to a pixel, and both
  // languages then take the same branch from the same integers.
  const vivid = vividTopShare > 0 && vividApplies(srcW, srcH, gridWidth, gridHeight);
  // One bucket per value `max − min` can take, reused across cells: the cell's weight by chroma, which turns
  // "the most colourful quarter" into an integer cutoff both languages reach by the same comparisons.
  const chromaWeight = vivid ? new Float64Array(256) : undefined;

  for (let cellY = 0; cellY < gridHeight; cellY++) {
    const ySrcStart = (cellY * srcH) / gridHeight;
    const ySrcEnd = ((cellY + 1) * srcH) / gridHeight;
    const yFirst = Math.max(0, Math.floor(ySrcStart));
    const yLast = Math.min(srcH - 1, Math.ceil(ySrcEnd) - 1);

    for (let cellX = 0; cellX < gridWidth; cellX++) {
      const xSrcStart = (cellX * srcW) / gridWidth;
      const xSrcEnd = ((cellX + 1) * srcW) / gridWidth;
      const xFirst = Math.max(0, Math.floor(xSrcStart));
      const xLast = Math.min(srcW - 1, Math.ceil(xSrcEnd) - 1);

      let sumR = 0;
      let sumG = 0;
      let sumB = 0;
      let sumWeight = 0;
      // The footprint's own area, so a cell clipped at the photo's border is judged on the part that exists.
      let sumArea = 0;

      for (let y = yFirst; y <= yLast; y++) {
        const yWeight = Math.min(y + 1, ySrcEnd) - Math.max(y, ySrcStart);
        if (yWeight <= 0) continue;
        for (let x = xFirst; x <= xLast; x++) {
          const xWeight = Math.min(x + 1, xSrcEnd) - Math.max(x, xSrcStart);
          if (xWeight <= 0) continue;
          const pixelIndex = (y * srcW + x) * 4;
          const alpha = data[pixelIndex + 3] / 255;
          const area = xWeight * yWeight;
          const weight = area * alpha;
          sumArea += area;
          sumR += SRGB_TO_LINEAR[data[pixelIndex]] * weight;
          sumG += SRGB_TO_LINEAR[data[pixelIndex + 1]] * weight;
          sumB += SRGB_TO_LINEAR[data[pixelIndex + 2]] * weight;
          sumWeight += weight;
        }
      }

      const i = cellY * gridWidth + cellX;
      coverage[i] = sumArea > 0 ? sumWeight / sumArea : 0;
      if (sumWeight > 0 && chromaWeight) {
        const mean: RGB = [linearToSrgb(sumR / sumWeight), linearToSrgb(sumG / sumWeight), linearToSrgb(sumB / sumWeight)];
        const vividRgb = vividCellColor(data, mean, srcW, xFirst, xLast, yFirst, yLast, xSrcStart, xSrcEnd, ySrcStart, ySrcEnd, sumWeight, vividTopShare, chromaWeight);
        out[i * 3] = vividRgb[0];
        out[i * 3 + 1] = vividRgb[1];
        out[i * 3 + 2] = vividRgb[2];
      } else if (sumWeight > 0) {
        out[i * 3] = linearToSrgb(sumR / sumWeight);
        out[i * 3 + 1] = linearToSrgb(sumG / sumWeight);
        out[i * 3 + 2] = linearToSrgb(sumB / sumWeight);
      } else {
        out[i * 3] = 255;
        out[i * 3 + 1] = 255;
        out[i * 3 + 2] = 255;
      }
    }
  }

  return { cells: { data: out, width: gridWidth, height: gridHeight }, coverage };
}

/**
 * One cell's Vivid colour: the area mean's lightness, with the chroma of the most colourful `topShare` of the cell's
 * weight (G-061, D211). The cutoff is an integer chroma-proxy value, reached by walking the weight histogram down
 * from the most colourful bucket, so both languages select the same pixels; every pixel at or above it is then
 * averaged in linear light with the same alpha-and-area weights the mean used. The recombination is gamut-mapped
 * rather than clipped, because clipping a channel turns an out-of-gamut colour into a different hue (D111).
 */
function vividCellColor(
  data: Uint8ClampedArray,
  mean: RGB,
  srcW: number,
  xFirst: number,
  xLast: number,
  yFirst: number,
  yLast: number,
  xSrcStart: number,
  xSrcEnd: number,
  ySrcStart: number,
  ySrcEnd: number,
  sumWeight: number,
  topShare: number,
  chromaWeight: Float64Array
): RGB {
  chromaWeight.fill(0);
  for (let y = yFirst; y <= yLast; y++) {
    const yWeight = Math.min(y + 1, ySrcEnd) - Math.max(y, ySrcStart);
    if (yWeight <= 0) continue;
    for (let x = xFirst; x <= xLast; x++) {
      const xWeight = Math.min(x + 1, xSrcEnd) - Math.max(x, xSrcStart);
      if (xWeight <= 0) continue;
      const pixelIndex = (y * srcW + x) * 4;
      chromaWeight[chromaProxy(data, pixelIndex)] += xWeight * yWeight * (data[pixelIndex + 3] / 255);
    }
  }

  // The most colourful bucket whose weight, with everything above it, first reaches the share asked for. Ties are
  // kept whole: a cell of one flat colour selects every pixel and is its own mean again.
  const wanted = sumWeight * topShare;
  let kept = 0;
  let cutoff = 0;
  for (let bucket = 255; bucket >= 0; bucket--) {
    kept += chromaWeight[bucket];
    if (kept >= wanted) {
      cutoff = bucket;
      break;
    }
  }

  let sumR = 0;
  let sumG = 0;
  let sumB = 0;
  let weight = 0;
  for (let y = yFirst; y <= yLast; y++) {
    const yWeight = Math.min(y + 1, ySrcEnd) - Math.max(y, ySrcStart);
    if (yWeight <= 0) continue;
    for (let x = xFirst; x <= xLast; x++) {
      const xWeight = Math.min(x + 1, xSrcEnd) - Math.max(x, xSrcStart);
      if (xWeight <= 0) continue;
      const pixelIndex = (y * srcW + x) * 4;
      if (chromaProxy(data, pixelIndex) < cutoff) continue;
      const w = xWeight * yWeight * (data[pixelIndex + 3] / 255);
      sumR += SRGB_TO_LINEAR[data[pixelIndex]] * w;
      sumG += SRGB_TO_LINEAR[data[pixelIndex + 1]] * w;
      sumB += SRGB_TO_LINEAR[data[pixelIndex + 2]] * w;
      weight += w;
    }
  }
  if (weight <= 0) return mean;

  const [, a, b] = rgbToOklab([linearToSrgb(sumR / weight), linearToSrgb(sumG / weight), linearToSrgb(sumB / weight)]);
  const [lightness] = rgbToOklab(mean);
  return oklabToRgbGamutMapped([lightness, a, b]);
}

File: rust/cs-core/src/downsample.rs
//! Port of `lib/pipeline/downsample.ts`.

use crate::color::{gamut_map_oklab_to_linear, linear_to_srgb, rgb_to_oklab, srgb_to_linear_table, Rgb};
use crate::jsmath;
use crate::Image;
use rayon::prelude::*;

/// `gridDimensionsFor`.
pub fn grid_dimensions_for(
    source_width: usize,
    source_height: usize,
    longer_side_stitches: f64,
) -> (usize, usize) {
    let longer = jsmath::max(1.0, jsmath::round(longer_side_stitches));
    let (sw, sh) = (source_width as f64, source_height as f64);
    if source_width >= source_height {
        let height = jsmath::max(1.0, jsmath::round((longer * sh) / sw));
        (longer as usize, height as usize)
    } else {
        let width = jsmath::max(1.0, jsmath::round((longer * sw) / sh));
        (width as usize, longer as usize)
    }
}

/// `MIN_CELL_COVERAGE`: a cell covered less than this is an empty stitch, not a colour (G-050).
pub const MIN_CELL_COVERAGE: f32 = 0.5;

/// `emptyCellMask`: which cells the photo barely covers, or `None` when it covers them all.
pub fn empty_cell_mask(coverage: &[f32]) -> Option<Vec<u8>> {
    let mut mask: Option<Vec<u8>> = None;
    for (i, &c) in coverage.iter().enumerate() {
        if c >= MIN_CELL_COVERAGE {
            continue;
        }
        mask.get_or_insert_with(|| vec![0u8; coverage.len()])[i] = 1;
    }
    mask
}

/// `downsampleToGrid`: the colours alone, for callers with no use for coverage.
pub fn downsample_to_grid(image: &Image, grid_width: usize, grid_height: usize) -> Vec<u8> {
    downsample_to_grid_with_coverage(image, grid_width, grid_height).0
}

/// Vivid (G-061): the share of each cell's footprint, by weight, whose chroma the cell keeps. 0 is off.
pub const VIVID_TOP_SHARE: f64 = 0.25;

/// Vivid needs a stitch to hold enough pixels for "its most colourful quarter" to be a real subpopulation rather
/// than a lucky pixel; below this it stands down and the cell is its plain area mean (D211).
pub const VIVID_MIN_PIXELS_PER_CELL: f64 = 24.0;

/// Whether Vivid can act on this photo at this chart size, which is what a pattern records rather than what was asked.
pub fn vivid_applies(
    source_width: usize,
    source_height: usize,
    grid_width: usize,
    grid_height: usize,
) -> bool {
    (source_width * source_height) as f64 / (grid_width * grid_height) as f64
        >= VIVID_MIN_PIXELS_PER_CELL
}

/// `max - min` of the sRGB bytes: a whole-number stand-in for chroma, so both languages select the same pixels.
fn chroma_proxy(data: &[u8], p: usize) -> usize {
    let (r, g, b) = (data[p], data[p + 1], data[p + 2]);
    (r.max(g).max(b) - r.min(g).min(b)) as usize
}

/// `downsampleToGridWithCoverage`: area- and alpha-weighted mean in linear light, one RGB triple per cell, plus the
/// alpha-weighted share of each cell's footprint that is actually there.
pub fn downsample_to_grid_with_coverage(
    image: &Image,
    grid_width: usize,
    grid_height: usize,
) -> (Vec<u8>, Vec<f32>) {
    downsample_to_grid_vivid(image, grid_width, grid_height, 0.0)
}

/// `downsampleToGridWithCoverage` with Vivid's share: above 0 a cell keeps its area-mean lightness and the chroma of
/// its most colourful `vivid_top_share` of weight, unless the photo gives a stitch too few pixels to tell (D211).
pub fn downsample_to_grid_vivid(
    image: &Image,
    grid_width: usize,
    grid_height: usize,
    vivid_top_share: f64,
) -> (Vec<u8>, Vec<f32>) {
    let table = srgb_to_linear_table();
    let (src_w, src_h) = (image.width, image.height);
    // A whole-image decision, not a per-cell one, so both languages take the same branch from the same integers.
    let vivid = vivid_top_share > 0.0 && vivid_applies(src_w, src_h, grid_width, grid_height);
    let data = &image.data;
    let mut out = vec![0u8; grid_width * grid_height * 3];
    let mut coverage = vec![0f32; grid_width * grid_height];

    // Each cell's sum runs in the same order on any thread count: rows of cells are independent.
    out.par_chunks_mut(grid_width * 3)
        .zip(coverage.par_chunks_mut(grid_width))
        .enumerate()
        .for_each(|(cell_y, (out, coverage))| {
            // One bucket per value `max - min` can take, reused across this row's cells.
            let mut chroma_weight = [0f64; 256];
            let y_start = (cell_y * src_h) as f64 / grid_height as f64;
            let y_end = ((cell_y + 1) * src_h) as f64 / grid_height as f64;
            let y_first = jsmath::max(0.0, y_start.floor()) as usize;
            let y_last = jsmath::min((src_h - 1) as f64, y_end.ceil() - 1.0) as i64;

            for cell_x in 0..grid_width {
                let x_start = (cell_x * src_w) as f64 / grid_width as f64;
                let x_end = ((cell_x + 1) * src_w) as f64 / grid_width as f64;
                let x_first = jsmath::max(0.0, x_start.floor()) as usize;
                let x_last = jsmath::min((src_w - 1) as f64, x_end.ceil() - 1.0) as i64;

                let mut sum_r = 0.0;
                let mut sum_g = 0.0;
                let mut sum_b = 0.0;
                let mut sum_weight = 0.0;
                // The footprint's own area, so a cell clipped at the photo's border is judged on the part that exists.
                let mut sum_area = 0.0;

                let mut y = y_first as i64;
                while y <= y_last {
                    let yf = y as f64;
                    let y_weight = jsmath::min(yf + 1.0, y_end) - jsmath::max(yf, y_start);
                    if y_weight > 0.0 {
                        let mut x = x_first as i64;
                        while x <= x_last {
                            let xf = x as f64;
                            let x_weight = jsmath::min(xf + 1.0, x_end) - jsmath::max(xf, x_start);
                            if x_weight > 0.0 {
                                let p = (y as usize * src_w + x as usize) * 4;
                                let alpha = data[p + 3] as f64 / 255.0;
                                let area = x_weight * y_weight;
                                let weight = area * alpha;
                                sum_area += area;
                                sum_r += table[data[p] as usize] * weight;
                                sum_g += table[data[p + 1] as usize] * weight;
                                sum_b += table[data[p + 2] as usize] * weight;
                                sum_weight += weight;
                            }
                            x += 1;
                        }
                    }
                    y += 1;
                }

                let i = cell_x * 3;
                coverage[cell_x] = if sum_area > 0.0 {
                    (sum_weight / sum_area) as f32
                } else {
                    0.0
                };
                if sum_weight > 0.0 && vivid {
                    let mean: Rgb = [
                        linear_to_srgb(sum_r / sum_weight),
                        linear_to_srgb(sum_g / sum_weight),
                        linear_to_srgb(sum_b / sum_weight),
                    ];
                    let rgb = vivid_cell_color(
                        data,
                        mean,
                        src_w,
                        table,
                        x_first,
                        x_last,
                        y_first,
                        y_last,
                        x_start,
                        x_end,
                        y_start,
                        y_end,
                        sum_weight,
                        vivid_top_share,
                        &mut chroma_weight,
                    );
                    out[i] = rgb[0];
                    out[i + 1] = rgb[1];
                    out[i + 2] = rgb[2];
                } else if sum_weight > 0.0 {
                    out[i] = linear_to_srgb(sum_r / sum_weight);
                    out[i + 1] = linear_to_srgb(sum_g / sum_weight);
                    out[i + 2] = linear_to_srgb(sum_b / sum_weight);
                } else {
                    out[i] = 255;
                    out[i + 1] = 255;
                    out[i + 2] = 255;
                }
            }
        });
    (out, coverage)
}

/// One cell's Vivid colour: the area mean's lightness with the chroma of its most colourful `top_share` of weight
/// (G-061, D211). Port of `vividCellColor`, in the same two passes and the same order, so the doubles match.
#[allow(clippy::too_many_arguments)]
fn vivid_cell_color(
    data: &[u8],
    mean: Rgb,
    src_w: usize,
    table: &[f64; 256],
    x_first: usize,
    x_last: i64,
    y_first: usize,
    y_last: i64,
    x_start: f64,
    x_end: f64,
    y_start: f64,
    y_end: f64,
    sum_weight: f64,
    top_share: f64,
    chroma_weight: &mut [f64; 256],
) -> Rgb {
    chroma_weight.fill(0.0);
    let mut y = y_first as i64;
    while y <= y_last {
        let yf = y as f64;
        let y_weight = jsmath::min(yf + 1.0, y_end) - jsmath::max(yf, y_start);
        if y_weight > 0.0 {
            let mut x = x_first as i64;
            while x <= x_last {
                let xf = x as f64;
                let x_weight = jsmath::min(xf + 1.0, x_end) - jsmath::max(xf, x_start);
                if x_weight > 0.0 {
                    let p = (y as usize * src_w + x as usize) * 4;
                    chroma_weight[chroma_proxy(data, p)] +=
                        x_weight * y_weight * (data[p + 3] as f64 / 255.0);
                }
                x += 1;
            }
        }
        y += 1;
    }

    // The most colourful bucket whose weight, with everything above it, first reaches the share asked for.
    let wanted = sum_weight * top_share;
    let mut kept = 0.0;
    let mut cutoff = 0usize;
    for bucket in (0..=255usize).rev() {
        kept += chroma_weight[bucket];
        if kept >= wanted {
            cutoff = bucket;
            break;
        }
    }

    let mut sum_r = 0.0;
    let mut sum_g = 0.0;
    let mut sum_b = 0.0;
    let mut weight = 0.0;
    let mut y = y_first as i64;
    while y <= y_last {
        let yf = y as f64;
        let y_weight = jsmath::min(yf + 1.0, y_end) - jsmath::max(yf, y_start);
        if y_weight > 0.0 {
            let mut x = x_first as i64;
            while x <= x_last {
                let xf = x as f64;
                let x_weight = jsmath::min(xf + 1.0, x_end) - jsmath::max(xf, x_start);
                if x_weight > 0.0 {
                    let p = (y as usize * src_w + x as usize) * 4;
                    if chroma_proxy(data, p) >= cutoff {
                        let w = x_weight * y_weight * (data[p + 3] as f64 / 255.0);
                        sum_r += table[data[p] as usize] * w;
                        sum_g += table[data[p + 1] as usize] * w;
                        sum_b += table[data[p + 2] as usize] * w;
                        weight += w;
                    }
                }
                x += 1;
            }
        }
        y += 1;
    }
    if weight <= 0.0 {
        return mean;
    }

    let top = rgb_to_oklab([
        linear_to_srgb(sum_r / weight),
        linear_to_srgb(sum_g / weight),
        linear_to_srgb(sum_b / weight),
    ]);
    let lightness = rgb_to_oklab(mean)[0];
    let lin = gamut_map_oklab_to_linear(lightness, top[1], top[2]);
    [
        linear_to_srgb(lin[0]),
        linear_to_srgb(lin[1]),
        linear_to_srgb(lin[2]),
    ]
}
